paymentsheet/src/main/java/com/stripe/android/paymentelement/embedded/EmbeddedRowSelectionImmediateActionHandler.kt

```diff
@@ -0,0 +1,117 @@
+package com.stripe.android.paymentelement.embedded
+
+import androidx.lifecycle.SavedStateHandle
+import com.stripe.android.core.injection.ViewModelScope
+import com.stripe.android.paymentelement.ExperimentalEmbeddedPaymentElementApi
+import com.stripe.android.paymentelement.embedded.content.DefaultPaymentOptionDisplayDataHolder
+import com.stripe.android.paymentelement.embedded.form.FormResult
+import com.stripe.android.paymentsheet.analytics.code
+import com.stripe.android.paymentsheet.model.PaymentSelection
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+
+internal typealias InternalRowSelectionCallback = () -> Unit
+
+@Singleton
+internal interface EmbeddedRowSelectionImmediateActionHandler {
+    fun prepareRowSelectionCallbackFromForm(formResult: FormResult.Complete)
+    fun prepareRowSelectionCallbackNonFormRows(selectionCode: String)
+    fun prepareRowSelectionCallbackSavedPaymentRow(updatedSelection: PaymentSelection.Saved)
+}
+
+@OptIn(ExperimentalEmbeddedPaymentElementApi::class)
+@Singleton
+internal class DefaultEmbeddedRowSelectionImmediateActionHandler @Inject constructor(
+    @ViewModelScope private val coroutineScope: CoroutineScope,
+    private val savedStateHandle: SavedStateHandle,
+    private val selectionHolder: EmbeddedSelectionHolder,
+    private val internalRowSelectionCallback: Provider<InternalRowSelectionCallback?>,
+    private val paymentOptionDisplayDataHolder: DefaultPaymentOptionDisplayDataHolder,
+) : EmbeddedRowSelectionImmediateActionHandler {
+    val selection = selectionHolder.selection
+
+    private val shouldInvokeRowSelectionCallback: StateFlow<Boolean> = savedStateHandle.getStateFlow(
+        key = EMBEDDED_ROW_SELECTION_SHOULD_INVOKE_KEY,
+        initialValue = false
+    )
+    private val promisedSelectionId: StateFlow<String?> = savedStateHandle.getStateFlow(
+        key = EMBEDDED_ROW_SELECTION_PROMISED_ID_KEY,
+        initialValue = null
+    )
+
+    init {
+        coroutineScope.launch {
+            paymentOptionDisplayDataHolder.paymentOption.collect { paymentOption ->
```

Comment:
I think we need to drive this from the UI or the interactor handleViewAction, I don't see a way for this to work being driven from the data model that can change for other reasons (user removes a saved payment method shouldn't trigger this).